src/Catalyst.Core.Modules.Dfs/CoreApi/DnsApi.cs

```diff
@@ -0,0 +1,160 @@
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Catalyst.Node. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#endregion
+
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Catalyst.Abstractions.Dfs.CoreApi;
+using Lib.P2P;
+using Makaretu.Dns;
+
+namespace Catalyst.Core.Modules.Dfs.CoreApi
+{
+    internal sealed class DnsApi : IDnsApi
+    {
+        private readonly IDnsClient _dnsClient;
+
+        public DnsApi(IDnsClient dnsClient) { _dnsClient = dnsClient; }
+
+        public async Task<string>
+            ResolveNameAsync(string name,
+                bool recursive = false,
+                bool nocache = false,
+                CancellationToken cancel = default)
+        {
+            do
+            {
+                if (name.StartsWith("/ipns/"))
+                {
+                    name = name.Substring(6);
+                }
+
+                var parts = name.Split('/').Where(p => p.Length > 0).ToArray();
+                if (parts.Length == 0)
+                    throw new ArgumentException($"Cannot resolve '{name}'.");
+                if (IsDomainName(parts[0]))
+                {
+                    name = await ResolveAsync(parts[0], recursive, cancel).ConfigureAwait(false);
+                }
+                else
+                {
+                    throw new NotImplementedException("Resolving IPNS is not yet implemented.");
+                }
+
+                if (parts.Length > 1)
+                {
+                    name = name + "/" + string.Join("/", parts, 1, parts.Length - 1);
+                }
+            } while (recursive && !name.StartsWith("/ipfs/"));
+
+            return name;
+        }
+
+        public async Task<string> ResolveAsync(string name, bool recursive = false, CancellationToken cancel = default)
+        {
+            // Find the TXT dnslink in either <name> or _dnslink.<name>.
+            string link = null;
+            using (var cts = CancellationTokenSource.CreateLinkedTokenSource(cancel))
+            {
+                try
+                {
+                    var attempts = new[]
+                    {
+                        FindAsync(name, cts.Token),
+                        FindAsync("_dnslink." + name, cts.Token)
+                    };
+                    link = await TaskHelper.WhenAnyResultAsync(attempts, cancel).ConfigureAwait(false);
+                    cts.Cancel();
+                }
+                catch (Exception e)
+                {
+                    throw new NotSupportedException($"Cannot resolve '{name}'.", e);
+                }
+            }
+
+            if (!recursive || link.StartsWith("/ipfs/"))
+                return link;
+
+            if (link.StartsWith("/ipns/"))
+            {
+                return await ResolveNameAsync(link, recursive, false, cancel).ConfigureAwait(false);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this default value assigned to parameter 'nocache'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051852)